feat: Update to latest sentry-native 0.4.0

Swatinem opened the pull request:
This updates the submodule to https://github.com/getsentry/sentry-native/pull/335, which should land hopefully soon and which will then be released as `0.4.0`.

Just putting this up so we can run this through your excellent testsuite as well ;-) (which I do have some problems getting to work locally however -_-)

Since we have done quite some bulletproofing, it might be worth removing any kind of locking you do around the global options / API. Can you point me in the right direction there maybe?

Swatinem committed:
update submodule and sys definitions

Swatinem committed:
update crate for API changes

Swatinem commented:
So far I have only tested this on my local windows. I ran into cmake build problems related to the path limit, but working around those, I ran into this in the `threaded_stress` test:

```
[sentry] DEBUG starting transport
[sentry] WARN `WinHttpOpen` failed
[sentry] WARN failed to initialize transport
[sentry] WARN `sentry_init` failed
```

propagating the transport startup failure to `sentry_init` is new in this version, but I have newer seen `WinHttpOpen` failing before.

Swatinem commented:
Also, I do get a `Error: HTTP status client error (403 Forbidden) for url (https://sentry.io/api/0/projects/)` when running locally, but I have set the `SENTRY_DSN` correctly, and the `SENTRY_TOKEN` to my personal API token, right?

daxpedda commented:
https://github.com/daxpedda/sentry-contrib-native/blob/dbc7c2262f139b3629af7619d08874335e3db5d5/src/options.rs#L27

Just remove this line, this in turn will invalidate `global_read`, `global_write`, which you can remove too, including all references to it.

daxpedda commented:
Is there a check now to not call `sentry_init` twice? Because `GLOBAL_LOCK` was handling that too. If this is not checked from `sentry_native`s side, make sure to replace it with something appropriate.

Swatinem commented:
> Is there a check now to not call `sentry_init` twice? Because `GLOBAL_LOCK` was handling that too.

well not directly, but we now shut down everything correctly when you call `sentry_init` a second time. note though that the options passed are still moved, so you will lose ownership.

daxpedda commented:
Oh, so now it's allowed to call `sentry_init` twice?

daxpedda commented:
https://github.com/getsentry/sentry-native/pull/341/files#r465647762 this question has come up.

daxpedda committed:
Continue update.

- Add `cfg(crashpad)` to ease checks.
- Changed default backend for Linux to Crashpad.
- Changed default transport for Android to Curl.
- Merge doc changes from `sentry.h`.
- Changed `BEFORE_SEND` and `LOGGER` behaviour to be deallocated on `shutdown` or on failed `Options::init`.
- Remove locking behaviour, thread safety is handled upstream.
- Keep locking behaviour in `shutdown` to enable proper deallocation.
- Changed logger to use `userdata` by making a new trait: `Logger`.
- Changed `set_transport` `startup` to return `Result`.
- Expand logger test to new functionality.
- Reenable attachment test in `crash` test.
- Improve attachment decoding in tests.
- Updated CHANGELOG.
- Updated README.

daxpedda committed:
Fix clippy.

daxpedda committed:
Fix clippy.

daxpedda commented:
I removed all the thread safety checks. With `Options` everything seems to work, but other functions are still not working.
You can try it locally with `cargo test --features test --lib threaded_stress -- --exact`. Just comment out everything in `src/lib.rs:591-672` except the functions you want to test.

On Linux I managed to boil it down to these functions:
- `set_extra`: fails completely on it's own, it's a bit hard to trigger, sometimes I have to run the test 20 times
- `set_context`: fails only together with `remove_context`

Both always give me the same error:
```
[xxx:ERROR scoped_ptrace_attach.cc:45] ptrace: No such process (3)
```

All other functions in `threaded_stress` together seem to work fine.

daxpedda committed:
Update CHANGELOG.

Swatinem commented:
This error is also coming from crashpad. Can you reproduce these failures without it?
Crashpad is a bit annoying as it can only be started once really. So there is no real way to shut it down once its started.

daxpedda commented:
> This error is also coming from crashpad. Can you reproduce these failures without it?

No, no errors if I use just breakpad.
> Crashpad is a bit annoying as it can only be started once really. So there is no real way to shut it down once its started.

I didn't look at the code thats producing the error, but I'm not starting or shutting down multiple times, I'm only executing this one test.

In any case, is the Crashpad backend for Linux considered experimental? Should I switch back to defaulting to Breakpad?
Should I file an issue upstream?

Swatinem commented:
> In any case, is the Crashpad backend for Linux considered experimental?

Not really, but it is brand new, so might be a bit untested.

> Should I switch back to defaulting to Breakpad?

Since that is what our own default is, it might be a good idea, yes.

I would have to look at all that code in detail, I have no idea what it does yet myself.

daxpedda commented:
> > Should I switch back to defaulting to Breakpad?
> 
> Since that is what our own default is, it might be a good idea, yes.

It seemed appropriate to me to switch the default to Crashpad, because Breakpad doesn't handle certain crashes, which I felt defeated the purpose of this library. Also https://github.com/daxpedda/sentry-contrib-native/issues/8 is a thing.

But yeah, I will switch it back to Breakpad.
> I would have to look at all that code in detail, I have no idea what it does yet myself.

I will also try to take a look at it.

daxpedda commented:
Another error that showed up on Crashpad on Linux:
```
[sentry] INFO started crashpad client handler
[sentry] INFO flushing session and queue before crashpad handler
[sentry] DEBUG sending envelope
[sentry] DEBUG serializing envelope into buffer
[sentry] INFO handing control over to crashpad
[xxx:ERROR elf_dynamic_array_reader.h:61] tag not found
```
It appears in the log after an application has crashed.

EDIT: I forgot to mention, that I couldn't find any real effect on the outcome. Even though there was an error, the upload was successful.

Swatinem commented:
@daxpedda There were indeed some thread-safety issues related to scope flush callbacks, those should be fixed here:
https://github.com/getsentry/sentry-native/pull/348

That should take care of the `set_extra` and `set_context` issues. And yes, I think using the crashpad backend introduced a little delay here that triggered the threadsafety issue:

1. ` set_context` and friends change the scope while locked, triggering a flush
2. the flush is propagated to the crashpad backend, which does its own locking so should be safe.
3. the session which is attached to the scope is also flushed while *not* holding the lock, which is probably triggering the crashes you see.

Again, thank you so so much for creating such an amazing testsuite, which has by now lead to a ton of fixes on our end! ❤️

Swatinem commented:
Also, those logs you get definitely come from crashpad, but I don’t know why they happen.
This is pure speculation on my part, but it might be possible that crashpad tries to capture multiple crashes at the same time happening in different threads?

Also, I think this specific issue is triggered by calling all the global functions concurrently with `sentry_shutdown`, am I correct?

daxpedda commented:
> Also, those logs you get definitely come from crashpad, but I don’t know why they happen.
> This is pure speculation on my part, but it might be possible that crashpad tries to capture multiple crashes at the same time happening in different threads?
> 
> Also, I think this specific issue is triggered by calling all the global functions concurrently with `sentry_shutdown`, am I correct?

`[xxx:ERROR elf_dynamic_array_reader.h:61] tag not found` if you mean this, it appears in the integration tests, which are run completely in serial and `shutdown` is only called after everything is done.

daxpedda committed:
Update to `sentry-native` to 0.4.0.

- Update `set_http_proxy`.
- Temporary fix for `threaded_stress`.
- Add `Options::set_auto_session_tracking` and `Options::auto_session_tracking`.
- Expanded options tests.

daxpedda committed:
Improve documentation and update CHANGELOG.

daxpedda commented:
As discussed on Discord, the thread safety issue isn't entirely fixed, but it is in some cases much more spurios, in others it can still be triggered "easily".

I will wait a bit and see if this can be resolved, if not I will re-implement some thread safety guards again and merge this.

Otherwise this is good to go.

daxpedda commented:
Now that it's working, I also intend to leave Crashpad as the default backend for Linux for the reasons explained above, unless there are any objections?

Will also release a 0.1.0 version on crates.io right after.

Swatinem commented:
>  Crashpad as the default backend for Linux

go for it!

daxpedda committed:
Re-introduced global lock.

Improved locking mechanism for `Options::init`.

daxpedda committed:
Fix clippy.
